### DISCUSSION TITLE
How to install local package

### DISCUSSION THREAD

**Discussion by the-sean-c** [pdm-project/pdm#2349] [Q&A] [answered]
Hi, I'm trying to locally install a package in the src folder. At the simplest, this would look something like the `src/example_package` that pdm makes by default when you run `pdm init`. However, if I try to install that package, using `pdm add ./src/example_package/`, then I get this error:

```
[RequirementError]: The local path 'src/example_package' is not installable.
```

This seems to me to be the way it should work based on the [docs](https://pdm.fming.dev/latest/usage/dependency/#local-dependencies). What am I doing wrong?

**Comment by pawamoy** [comment, marked as answer]
There is no `pyproject.toml` in `src/example_package`, so no Python installation tool would know how to install it. If you want to list a local package as a dependency of your project, you can indeed use `pdm add`, but you have to point it at the root folder which contains `pyproject.toml`. Also, you don't need to add the current project as a dependency of itself! In your case I believe you can just run `pdm install`.

**Reply by the-sean-c** [reply]
Hmm. I tried `pdm install`, but it gave me `[KeyError]: 'name'`.

Instead, I deleted the project and started again with `pdm init`, this time when it asked "Is the project a library that is installable?", I said yes. now after `pdm install` pdm picks up the packages under src automatically. This is like poetry, which is what I expected to happen.

Looks like that solved it, thanks!

**Comment by Hopsakee** [comment]
I could still use some help here. I'm afraid I'm missing some very basic knowledge.

My projectfolder has this structure:

```
PDMtest/
    ├── src/
    │   └── __init__.py
    │   └── print_helper.py
    └── tests/
    │   └── runner_tests.py
    └── runner_root.py
```

`runner_tests.py` and `runner_root.py` both look like this:

```
from src import print_helper

print_helper.all_of_it()
```

Now. The file `runner_tests.py` gives an import error:
```
ModuleNotFoundError: No module named 'src'
```

While the file `runner_root.py` runs just fine.

Can someone explain what I'm missing?

I tried:
- After `pdm init` => "Is the project a library that is installable?", YES
- pdm add -e ./src
- pdm add ./src

And some other pdm add commands.

The 'pdm add' like commands give errors like this:
```
[PdmUsageError]: Cannot add editables to the default or optional dependency group
```

Or
```
[RequirementError]: The local path 'src' is not installable.
```

I also tried adding to my pyproject.toml

```
dependencies = [
    "file:///${PROJECT_ROOT}/src",
    ]
```

This gives the [RequirementError] when I do `pdm install`.

What does work is adding the `src` folder to my pythonpath like this in the `runner_tests.py` file:

```
import sys
import os
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))
```

But to me that looks like a workaround instead of a solution. Are there any more robust and pythonic PDM ways to use the code that is located in a folder from within a file that is itself located in another folder?

**Reply by pawamoy** [reply]
As mentioned above, you don't need to add the sources as a dependency of the project itself (it doesn't make sense anyway to try and add `src` as a dependency).

When using the src-layout, you are supposed to have this file tree:

```
project/
    src/
        your_package/
            __init__.py
```

You are missing `your_package` in your tree.

Surely you don't want to name your package "src", right :smile: 

Then, in your runner modules, import from `your_package`:

```python
from your_package import print_helper
```

Finally, to run either of your runner scripts, use `pdm run python runner_root.py`, or `pdm run python tests/runner_tests.py` (but in the latter case I suppose you're rather using something like pytest, so that rather be `pdm run pytest`). PDM takes care of finding your package in `src` and putting it in `PYTHONPATH`/`sys.path` for you.

**Reply by pawamoy** [reply]
More info on src-layout vs. flat-layout: https://packaging.python.org/en/latest/discussions/src-layout-vs-flat-layout/.

**Reply by Hopsakee** [reply]
@pawamoy Thanks. This really helped, in combination with the earlier advise to `pdm init` with "installable library".

Indeed, I wasn't planning on naming my package 'src'. 😄 But the fact that I suggested that, in my attempt to make my example as simple as possible, already suggests I couldn't wrap my head around what I was doing exactly. 😆 Your reference to the src-layout vs flat-layout was very helpful. Thanks.

**Reply by pawamoy** [reply]
Glad I could help! The different layouts cause a lot of confusion, you're not the only one ^^
